File: module/module-nms-util/src/main/java/taboolib/module/nms/ItemTag.java
package taboolib.module.nms;

import com.google.common.collect.Maps;
import com.google.gson.*;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.function.BiConsumer;
import java.util.function.BiFunction;
import java.util.function.Function;

/**
 * 物品 NBT 结构映射类
 *
 * @author 坏黑
 * @since 2019-05-24 17:44
 */
public class ItemTag extends ItemTagData implements Map<String, ItemTagData> {

    private final Map<String, ItemTagData> value = Maps.newConcurrentMap();

    public ItemTag() {
        super(0);
        this.type = ItemTagType.COMPOUND;
        this.data = this;
    }

    public void saveTo(ItemStack item) {
        item.setItemMeta(NMSKt.setItemTag(item, this).getItemMeta());
    }

    public String toJson() {
        return ItemTagSerializer.INSTANCE.serializeData(this).toString();
    }

    public String toJsonFormatted() {
        return new GsonBuilder().setPrettyPrinting().create().toJson(ItemTagSerializer.INSTANCE.serializeTag(this));
    }

    @Override
    public String toJsonSimplified() {
        return toJsonSimplified(0);
    }

    @Override
    public String toJsonSimplified(int index) {
        StringBuilder builder = new StringBuilder();
        builder.append("{\n");
        value.forEach((k, v) -> builder.append(copy("  ", index + 1))
                .append("\"")
                .append(k)
                .append("\"")
                .append(": ")
                .append(v.toJsonSimplified(index + 1))
                .append("\n"));
        builder.append(copy("  ", index)).append("}");
        return builder.toString();
    }

    public static ItemTag fromJson(String json) {
        return fromJson(new JsonParser().parse(json)).asCompound();
    }

    public static ItemTagData fromJson(JsonElement element) {
        return ItemTagSerializer.INSTANCE.deserializeData(element);
    }

    @Deprecated
    public String toLegacyJson() {
        return new Gson().toJson(this);
    }

    @Deprecated
    public static ItemTag fromLegacyJson(String json) {
        return fromLegacyJson(new JsonParser().parse(json)).asCompound();
    }

    @Deprecated
    public static ItemTagData fromLegacyJson(JsonElement element) {
        if (element instanceof JsonObject) {
            JsonObject json = (JsonObject) element;
            // base
            if (json.has("type") && json.has("data") && json.entrySet().size() == 2) {
                switch (ItemTagType.parse(json.get("type").getAsString())) {
                    case BYTE:
                        return new ItemTagData(json.get("data").getAsByte());
                    case SHORT:
                        return new ItemTagData(json.get("data").getAsShort());
                    case INT:
                        return new ItemTagData(json.get("data").getAsInt());
                    case LONG:
                        return new ItemTagData(json.get("data").getAsLong());
                    case FLOAT:
                        return new ItemTagData(json.get("data").getAsFloat());
                    case DOUBLE:
                        return new ItemTagData(json.get("data").getAsDouble());
                    case STRING:
                        return new ItemTagData(json.get("data").getAsString());
                    case BYTE_ARRAY: {
                        JsonArray array = json.get("data").getAsJsonArray();
                        byte[] bytes = new byte[array.size()];
                        for (int i = 0; i < array.size(); i++) {
                            bytes[i] = array.get(i).getAsByte();
                        }
                        return new ItemTagData(bytes);
                    }
                    case INT_ARRAY: {
                        JsonArray array = json.get("data").getAsJsonArray();
                        int[] ints = new int[array.size()];
                        for (int i = 0; i < array.size(); i++) {
                            ints[i] = array.get(i).getAsInt();
                        }
                        return new ItemTagData(ints);
                    }
                    default:
                        return new ItemTagData("error: " + element);
                }
            }
            // compound
            else {
                ItemTag compound = new ItemTag();
                for (Entry<String, JsonElement> elementEntry : json.entrySet()) {
                    compound.put(elementEntry.getKey(), fromLegacyJson(elementEntry.getValue()));
                }
                return compound;
            }
        }
        // list
        else if (element instanceof JsonArray) {
            ItemTagList list = new ItemTagList();
            for (JsonElement jsonElement : (JsonArray) element) {
                list.add(fromLegacyJson(jsonElement));
            }
            return list;
        }
        return new ItemTagData("error: " + element);
    }

    @Override
    public int size() {
        return value.size();
    }

    @Override
    public boolean isEmpty() {
        return value.isEmpty();
    }

    @Override
    public boolean containsKey(Object key) {
        return value.containsKey(key);
    }

    @Override
    public boolean containsValue(Object value) {
        return this.value.containsValue(value);
    }

    @Override
    public ItemTagData get(Object key) {
        return value.get(key);
    }

    public ItemTagData getOrElse(String key, ItemTagData base) {
        return value.getOrDefault(key, base);
    }

    public ItemTagData getDeep(String key) {
        ItemTagData value = this;
        for (String keyStr : key.split("\\.")) {
            if ((value = value.asCompound().get(keyStr)) == null) {
                return null;
            }
        }
        return value;
    }

    public ItemTagData getDeepOrElse(String key, ItemTagData base) {
        return Optional.ofNullable(getDeep(key)).orElse(base);
    }

    @Override
    public ItemTagData put(String key, ItemTagData value) {
        return this.value.put(key, value);
    }

    public ItemTagData put(String key, Object value) {
        return this.value.put(key, ItemTagData.toNBT(value));
    }

    public ItemTagData putDeep(String key, Object value) {
        return putDeep(key, ItemTagData.toNBT(value));
    }

    public ItemTagData putDeep(String key, ItemTagData value) {
        ItemTagData compound = this, temp;
        String[] split = key.split("\\.");
        for (String node : split) {
            if (node.equalsIgnoreCase(split[split.length - 1])) {
                return ((ItemTag) compound).put(node, value);
            }
            if ((temp = compound.asCompound().get(node)) == null) {
                temp = new ItemTag();
                compound.asCompound().put(node, temp);
            }
            compound = temp;
        }
        return null;
    }

    public ItemTagData removeDeep(String key) {
        ItemTagData compound = this, temp;
        String[] split = key.split("\\.");
        for (String node : split) {
            if (node.equalsIgnoreCase(split[split.length - 1])) {
                return ((ItemTag) compound).remove(node);
            }
            if ((temp = compound.asCompound().get(node)) == null) {
                return null;
            }
            compound = temp;
        }
        return null;
    }

    @Override
    public ItemTagData remove(Object key) {
        return value.remove(key);
    }

    @SuppressWarnings("unchecked")
    @Override
    public void putAll(@NotNull Map m) {
        this.value.putAll(m);
    }

    @Override
    public void clear() {
        this.value.clear();
    }

    @SuppressWarnings({"rawtypes", "unchecked"})
    @NotNull
    @Override
    public Set keySet() {
        return this.value.keySet();
    }

    @SuppressWarnings({"rawtypes", "unchecked"})
    @NotNull
    @Override
    public Collection values() {
        return this.value.values();
    }

    @NotNull
    @Override
    public Set<Entry<String, ItemTagData>> entrySet() {
        return this.value.entrySet();
    }

    @Override
    public ItemTagData getOrDefault(Object key, ItemTagData defaultValue) {
        return this.value.getOrDefault(String.valueOf(key), defaultValue);
    }

    @Override
    public void forEach(BiConsumer<? super String, ? super ItemTagData> action) {
        this.value.forEach(action);
    }

    @Override
    public void replaceAll(BiFunction<? super String, ? super ItemTagData, ? extends ItemTagData> function) {
        this.value.replaceAll(function);
    }

    @Override
    public ItemTagData putIfAbsent(String key, ItemTagData value) {
        return this.value.putIfAbsent(key, value);
    }

    @Override
    public boolean remove(Object key, Object value) {
        return this.value.remove(key, value);
    }

    @Override
    public boolean replace(String key, ItemTagData oldValue, ItemTagData newValue) {
        return this.value.replace(key, oldValue, newValue);
    }

    @Override
    public ItemTagData replace(String key, ItemTagData value) {
        return this.value.replace(key, value);
    }

    @Override
    public ItemTagData computeIfAbsent(String key, @NotNull Function<? super String, ? extends ItemTagData> mappingFunction) {
        return this.value.computeIfAbsent(key, mappingFunction);
    }

    @Override
    public ItemTagData computeIfPresent(String key, @NotNull BiFunction<? super String, ? super ItemTagData, ? extends ItemTagData> remappingFunction) {
        return this.value.computeIfPresent(key, remappingFunction);
    }

    @Override
    public ItemTagData compute(String key, @NotNull BiFunction<? super String, ? super ItemTagData, ? extends ItemTagData> remappingFunction) {
        return this.value.compute(key, remappingFunction);
    }

    @Override
    public ItemTagData merge(String key, @NotNull ItemTagData value, @NotNull BiFunction<? super ItemTagData, ? super ItemTagData, ? extends ItemTagData> remappingFunction) {
        return this.value.merge(key, value, remappingFunction);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof ItemTag)) {
            return false;
        }
        ItemTag that = (ItemTag) o;
        return Objects.equals(value, that.value);
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), value);
    }

    @Override
    public String toString() {
        return NMS_UTILS.itemTagToString(this);
    }
}

File: module/module-nms-util/src/main/kotlin/taboolib/module/nms/ItemTag.kt
package taboolib.module.nms

import com.google.gson.*
import org.bukkit.inventory.ItemStack
import taboolib.module.nms.ItemTagSerializer.serializeData
import taboolib.module.nms.ItemTagSerializer.serializeTag
import java.util.concurrent.ConcurrentHashMap

/**
 * TabooLib
 * taboolib.module.nms.ItemTag
 *
 * @author 坏黑
 * @since 2023/8/9 01:40
 */
class ItemTag : ItemTagData, MutableMap<String, ItemTagData> {

    private val value = ConcurrentHashMap<String, ItemTagData>()

    constructor() : super(ItemTagType.COMPOUND, 0) {
        this.data = this
    }

    constructor(map: Map<String, ItemTagData>) : super(ItemTagType.COMPOUND, 0) {
        this.data = this
        this.value += map
    }

    /**
     * 将 [ItemTag] 写入物品
     */
    fun saveTo(item: ItemStack) {
        item.setItemMeta(item.setItemTag(this).itemMeta)
    }

    /**
     * 转换为 Json（可逆向）
     */
    fun toJson(): String {
        return serializeData(this).toString()
    }

    /**
     * 转换为格式化的 Json（可逆向）
     */
    fun toJsonFormatted(): String {
        return GsonBuilder().setPrettyPrinting().create().toJson(serializeTag(this))
    }

    override fun asCompound(): ItemTag {
        return this
    }

    override fun toJsonSimplified(): String {
        return toJsonSimplified(0)
    }

    override fun toJsonSimplified(index: Int): String {
        val builder = StringBuilder()
        builder.append("{\n")
        value.forEach { (k: String?, v: ItemTagData) ->
            builder.append("  ".repeat(index + 1))
                .append("\"")
                .append(k)
                .append("\"")
                .append(": ")
                .append(v.toJsonSimplified(index + 1))
                .append("\n")
        }
        builder.append("  ".repeat(index)).append("}")
        return builder.toString()
    }

    override val entries: MutableSet<MutableMap.MutableEntry<String, ItemTagData>>
        get() = value.entries

    override val keys: MutableSet<String>
        get() = value.keys

    override val size: Int
        get() = value.size

    override val values: MutableCollection<ItemTagData>
        get() = value.values

    override fun clear() {
        value.clear()
    }

    override fun isEmpty(): Boolean {
        return value.isEmpty()
    }

    override fun remove(key: String): ItemTagData? {
        return value.remove(key)
    }

    /**
     * 深度删除，以 "." 作为分层符
     */
    fun removeDeep(key: String): ItemTagData? {
        return if (key.contains('.')) {
            getDeepWith(key, false) { it.remove(key.substringAfterLast('.')) }
        } else {
            remove(key)
        }
    }

    override fun putAll(from: Map<out String, ItemTagData>) {
        value += from
    }

    override fun put(key: String, value: ItemTagData): ItemTagData? {
        return this.value.put(key, value)
    }

    fun put(key: String, value: Any): ItemTagData? {
        return this.value.put(key, toNBT(value))
    }

    /**
     * 深度写入，以 "." 作为分层符
     */
    fun putDeep(key: String, value: Any): ItemTagData? {
        return if (key.contains('.')) {
            getDeepWith(key, true) { it.put(key.substringAfterLast('.'), toNBT(value)) }
        } else {
            put(key, value)
        }
    }

    override fun get(key: String): ItemTagData? {
        return value[key]
    }

    fun getOrElse(key: String, base: ItemTagData): ItemTagData {
        return value.getOrDefault(key, base)
    }

    /**
     * 深度获取，以 "." 作为分层符
     */
    fun getDeep(key: String): ItemTagData? {
        return if (key.contains('.')) {
            getDeepWith(key, false) { it[key.substringAfterLast('.')] }
        } else {
            get(key)
        }
    }

    /**
     * 针对 getDeep, putDeep, removeDeep 的重复代码做出的优化
     */
    fun getDeepWith(key: String, create: Boolean, action: (ItemTag) -> ItemTagData?): ItemTagData? {
        val keys = key.split('.').dropLast(1)
        if (keys.isEmpty()) {
            return null
        }
        var find: ItemTag = this
        for (element in keys) {
            var next = find[element]
            if (next == null) {
                if (create) {
                    next = ItemTag()
                    find[element] = next
                } else {
                    return null
                }
            }
            if (next.type == ItemTagType.COMPOUND) {
                find = next.asCompound()
            } else {
                return null
            }
        }
        return action(find)
    }

    /**
     * 深度获取，以 "." 作为分层符，并支持默认值
     */
    fun getDeepOrElse(key: String, base: ItemTagData): ItemTagData {
        return getDeep(key) ?: base
    }

    override fun containsValue(value: ItemTagData): Boolean {
        return this.value.containsValue(value)
    }

    override fun containsKey(key: String): Boolean {
        return value.containsKey(key)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ItemTag) return false
        if (value != other.value) return false
        return true
    }

    override fun hashCode(): Int {
        return value.hashCode()
    }

    override fun toString(): String {
        return saveToString()
    }

    @Deprecated("不推荐使用", ReplaceWith("Gson().toJson(this)", "com.google.gson.Gson"))
    fun toLegacyJson(): String {
        return Gson().toJson(this)
    }

    companion object {

        @JvmStatic
        fun fromJson(json: String): ItemTag {
            return fromJson(JsonParser().parse(json)).asCompound()
        }

        @JvmStatic
        fun fromJson(element: JsonElement): ItemTagData {
            return ItemTagSerializer.deserializeData(element)
        }

        @JvmStatic
        fun fromLegacyJson(json: String): ItemTag {
            return fromLegacyJson(JsonParser().parse(json)).asCompound()
        }

        @JvmStatic
        fun fromLegacyJson(element: JsonElement): ItemTagData {
            return if (element is JsonObject) {
                val json = element.asJsonObject
                // 基本类型
                if (json.has("type") && json.has("data") && json.entrySet().size == 2) {
                    when (val type = ItemTagType.parse(json.get("type").asString)) {
                        ItemTagType.BYTE -> ItemTagData(json.get("data").asByte)
                        ItemTagType.SHORT -> ItemTagData(json.get("data").asShort)
                        ItemTagType.INT -> ItemTagData(json.get("data").asInt)
                        ItemTagType.LONG -> ItemTagData(json.get("data").asLong)
                        ItemTagType.FLOAT -> ItemTagData(json.get("data").asFloat)
                        ItemTagType.DOUBLE -> ItemTagData(json.get("data").asDouble)
                        ItemTagType.STRING -> ItemTagData(json.get("data").asString)
                        // byte 数组
                        ItemTagType.BYTE_ARRAY -> {
                            val array = json.get("data").asJsonArray
                            val bytes = ByteArray(array.size())
                            for (i in 0 until array.size()) {
                                bytes[i] = array.get(i).asByte
                            }
                            ItemTagData(bytes)
                        }
                        // int 数组
                        ItemTagType.INT_ARRAY -> {
                            val array = json.get("data").asJsonArray
                            val ints = IntArray(array.size())
                            for (i in 0 until array.size()) {
                                ints[i] = array.get(i).asInt
                            }
                            ItemTagData(ints)
                        }
                        // 不支持的类型
                        else -> error("Unsupported nbt type: $type")
                    }
                } else {
                    // 复合类型
                    val compound = ItemTag()
                    for ((key, value) in json.entrySet()) {
                        compound[key] = fromLegacyJson(value)
                    }
                    compound
                }
            } else if (element is JsonArray) {
                val list = ItemTagList()
                for (value in element) {
                    list.add(fromLegacyJson(value))
                }
                list
            } else {
                error("Not JsonObject")
            }
        }
    }
}